src/lib.rs

```diff
@@ -1,12 +1,1184 @@
+                out.write_u32::<LittleEndian>(val.len() as u32).ok();
+                for &(hi, lo) in val {
+                    out.write_u64::<LittleEndian>(lo).ok();
+                    out.write_u64::<LittleEndian>(hi).ok();
+                }
+            }
+            &AUINT128(ref val) => {
+                out.write_u32::<LittleEndian>(val.len() as u32).ok();
+                for &(hi, lo) in val {
+                    out.write_u64::<LittleEndian>(lo).ok();
+                    out.write_u64::<LittleEndian>(hi).ok();
+                }
+            }
+            &STR(ref val) => {
+                out.write_u8(((val.len() as u8) * 2) + 1).ok();
+                if val.len() > 0 {
+                    for e in val.encode_utf16() { out.write_u16::<LittleEndian>(e).ok(); }
+                    out.write_all(b"\0\0").ok();
+                }
+            }
+            _ => {}
+        }
+        out
+    }
+
+    pub fn read_type<T: PtpRead>(kind: u16, reader: &mut T) -> Result<PtpDataType, Error> {
+        use self::PtpDataType::*;
+        Ok(match kind {
+            // 0x0000 => UNDEF,
+            0x0001 => INT8(try!(reader.read_ptp_i8())),
+            0x0002 => UINT8(try!(reader.read_ptp_u8())),
+            0x0003 => INT16(try!(reader.read_ptp_i16())),
+            0x0004 => UINT16(try!(reader.read_ptp_u16())),
+            0x0005 => INT32(try!(reader.read_ptp_i32())),
+            0x0006 => UINT32(try!(reader.read_ptp_u32())),
+            0x0007 => INT64(try!(reader.read_ptp_i64())),
+            0x0008 => UINT64(try!(reader.read_ptp_u64())),
+            0x0009 => INT128(try!(reader.read_ptp_i128())),
+            0x000A => UINT128(try!(reader.read_ptp_u128())),
+            0x4001 => AINT8(try!(reader.read_ptp_i8_vec())),
+            0x4002 => AUINT8(try!(reader.read_ptp_u8_vec())),
+            0x4003 => AINT16(try!(reader.read_ptp_i16_vec())),
+            0x4004 => AUINT16(try!(reader.read_ptp_u16_vec())),
+            0x4005 => AINT32(try!(reader.read_ptp_i32_vec())),
+            0x4006 => AUINT32(try!(reader.read_ptp_u32_vec())),
+            0x4007 => AINT64(try!(reader.read_ptp_i64_vec())),
+            0x4008 => AUINT64(try!(reader.read_ptp_u64_vec())),
+            0x4009 => AINT128(try!(reader.read_ptp_i128_vec())),
+            0x400A => AUINT128(try!(reader.read_ptp_u128_vec())),
+            0xFFFF => STR(try!(reader.read_ptp_str())),
+            _ => UNDEF,
+        })
+    }
+}
+
+impl<'a> From<&'a str> for PtpDataType {
+    fn from(value: &'a str) -> Self {
+        PtpDataType::STR(value.to_owned())
+    }
+}
+
+impl<'a> From<String> for PtpDataType {
+    fn from(value: String) -> Self {
+        PtpDataType::STR(value)
+    }
+}
+
+impl<'a> From<i8> for PtpDataType {
+    fn from(value: i8) -> Self {
+        PtpDataType::INT8(value)
+    }
+}
+
+impl<'a> From<i16> for PtpDataType {
+    fn from(value: i16) -> Self {
+        PtpDataType::INT16(value)
+    }
+}
+
+impl<'a> From<u8> for PtpDataType {
```

Comment:
While we're doing this PR, can we populate i32 (and whatever other types we expect, u64, i64, String / Vec maybe)? Maybe just the ints are all that's needed.